usb: cdc_acm: Additional connection state checks

metratec-mkoehler opened the pull request:
Issues caused:
- On intense use I had packet losses (below 1 per 1000)
- Restarting the host device or resetting the USB controller caused the usb transfer to get stuck in about 30% of tests
- This includes issue #72786 opened by a collegue

Analysis of issues:
  I found multiple cases causing this isssues with wrong return values on cdc_acm_irq_is_pending(), cdc_acm_irq_tx_ready() etc causing my logic to get stuck

Changes:
- usb_transfer(): k_work_submit_to_queue() return value was ignored -> calling function tx_work_handler() dropped data even though it was
   not put to workqueue
- tx_work_handler() put dev_data->tx_ready to false even if usb_transfer() failed but without the transfer done callback it was never reset to true so TX got stuck
- cdc_acm_irq_tx_ready() did not check configured and suspended states This caused the device to show as ready even though it was not configured after a host reset
- cdc_acm_irq_rx_ready() did not check if irq is enabled
- cdc_acm_irq_is_pending() had the same issues as its just a test on cdc_acm_irq_rx_ready() or cdc_acm_irq_tx_ready() (the description says so, too) -> Replaced it by a call of both functions for a fast and consistent fix

Tests:
- The host reset issue was tested with a Intel NUC with Windows 11 in an endless reset loop, about 2 times per minute, for 12 hours. The issue no longer applies there
- The host reset issue was also tested on a Linux XUbuntu host with a pyusb device reset script. The issue was tested only manually (at 50% success that was good enough) and does not apply any longer
- Data loss was tested with an echo device and a data generating and checking script. It was also tested with a device generating serialized data (counting up).
- Both showed data loss (about 1 byte per GB, in blocks of multiples of 8). For both test cases the issue no longer applied (tested with more than 1 TB of data) Fixes: #72786

Considerations:
- Changes in cdc_acm.c should not have impact on other code. Return values may show an error that was missed before
- Change in usb_transfer.c may influence more devices The new return value should be in an expected range (0 or negative) so I expect it to work with all other components as well

metratec-mkoehler commented:
Failing due to issue https://github.com/zephyrproject-rtos/zephyr/issues/73846

metratec-mkoehler commented:
https://github.com/zephyrproject-rtos/zephyr/pull/73838

Needed

josuah commented on subsys/usb/device/usb_transfer.c:
```diff
@@ -251,7 +251,8 @@ int usb_transfer(uint8_t ep, uint8_t *data, size_t dlen, unsigned int flags,
 		/* start writing first chunk */
-		k_work_submit_to_queue(&USB_WORK_Q, &trans->work);
+		ret = k_work_submit_to_queue(&USB_WORK_Q, &trans->work);
+		ret = MAX(0, ret);
```
I often mix min() an max(): `MAX(0, ret)` would take the greatest value between `0` and `ret`, so if `ret` is negative (i.e. error), and only pick 0, 1 or 2 as described here:
https://docs.zephyrproject.org/latest/kernel/services/threads/workqueue.html#c.k_work_submit_to_queue

Is that what is intended?
Thanks

metratec-mkoehler replied:
Thanks for the feedback. 
You are totally right, this is an error. Unsurprisingly the work queue never caused an issue in tests.
Fix incoming.

tmon-nordic commented on subsys/usb/device/usb_transfer.c:
```diff
@@ -251,7 +251,8 @@ int usb_transfer(uint8_t ep, uint8_t *data, size_t dlen, unsigned int flags,
 	if (flags & USB_TRANS_WRITE) {
 		/* start writing first chunk */
-		k_work_submit_to_queue(&USB_WORK_Q, &trans->work);
+		ret = k_work_submit_to_queue(&USB_WORK_Q, &trans->work);
```
Please split this to separate commit with its own commit message explaining what exactly it does.

metratec-mkoehler replied:
Done

tmon-nordic reviewed (changes requested):
Looks good. Please separate the CDC ACM changes and generic transfer handling changes - while together these do solve the same observed issue, they are really different root causes leading to the same symphtoms.

tmon-nordic commented:
You need to reword the commit messages. For instance the
```
- Change in usb_transfer.c may influence more devices
  The new return value should be in an expected range (0 or negative)
  so I expect it to work with all other components as well
```
is no longer relevant in the first commit.

The commit message format looks like originating from very formal process.

The second commit fails to really state what the change achieves and is mostly a copy of the first commit.

metratec-mkoehler commented:
The commit message format originates from
https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribute-guidelines
> 
    what the change does,
    why you chose that approach,
    what assumptions were made, and
    how you know it works – for example, which tests you ran.

But I will save me time in the future keeping things minimal instead. Change will come soon.

jfischer-no commented on subsys/usb/device/class/cdc_acm.c:
```diff
@@ -288,10 +287,13 @@ static void tx_work_handler(struct k_work *work)
-		     cdc_acm_write_cb, dev_data);
-
-	ring_buf_get_finish(dev_data->tx_ringbuf, len);
+	ret = usb_transfer(ep, data, len, USB_TRANS_WRITE, cdc_acm_write_cb, dev_data);
+	if (ret < 0) {
+		(void)ring_buf_get_finish(dev_data->tx_ringbuf, 0);
+		return;
+	}
+	dev_data->tx_ready = false;
+	(void)ring_buf_get_finish(dev_data->tx_ringbuf, len);
```
```c
	ret = usb_transfer(ep, data, len, USB_TRANS_WRITE,
			   cdc_acm_write_cb, dev_data);
	if (ret != 0) {
		(void)ring_buf_get_finish(dev_data->tx_ringbuf, 0);
		return;
	}

	dev_data->tx_ready = false;
	(void)ring_buf_get_finish(dev_data->tx_ringbuf, len);
```

metratec-mkoehler replied:
The line break is according to [the new formatting rule](https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style)  (100 chars per line).

metratec-mkoehler replied:
Also, ret < 0 is more robust to API changes than ret != 0 as giving back the number of finished bytes might be a future way to implement. If you want a security against an unexpected positive return value I would be fine with adding an
```
 __ASSERT(ret <= 0, "Unexpected return value");
```

jfischer-no replied:
> The line break is according to [the new formatting rule](https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style) (100 chars per line).

Sure ("The line length is 100 columns or fewer"), but I am asking you to change it in a minimal way without unnecessary reformatting.



> Also, ret < 0 is more robust to API changes than ret != 0 as giving back the number of finished bytes might be a future way to implement. If you want a security against an unexpected positive return value I would be fine with adding an
> 
> ```
>  __ASSERT(ret <= 0, "Unexpected return value");
> ```

No, there is no future for `usb_transfer()`, and I am asking for exactly that
```c
	if (ret != 0) {
		(void)ring_buf_get_finish(dev_data->tx_ringbuf, 0);
		return;
	}
```

metratec-mkoehler replied:
I will change it

metratec-mkoehler replied:
Changed. Also reordered the commits as the fix on cdc acm depends on the usb_transfer() fix so first transfer, then the cdc acm. The second one now states correctly that it fixes the issue.

jfischer-no commented on subsys/usb/device/class/cdc_acm.c:
```diff
@@ -621,13 +623,9 @@ static void cdc_acm_irq_tx_disable(const struct device *dev)
-	}
-
-	return 0;
+	return data->tx_irq_ena && data->configured && data->tx_ready && !data->suspended;
```
I do not think it is right, whether configured or not, UART device itself can fill data into FIFO, it will also break established expectations.

metratec-mkoehler replied:
I think it is right.

cdc_acm_fifo_fill() will not accept data. 

```
	if (!dev_data->configured || dev_data->suspended) {
		LOG_WRN("Device not configured or suspended, drop %d bytes",
			len);
		return 0;
	}
```

This cause an inifinte loop when using cdc_acm_irq_tx_ready() as in
zephyr/samples/subsys/usb/cdc_acm/src/main.c interrupt_handler() and many other projects.

jfischer-no replied:
cdc_acm_irq_tx_ready() is implementation of uart_irq_tx_ready(), where the descriptions is "Check if UART TX buffer can accept a new char", `TX buffer`, it is not `Check if UART controller can transfer a new char`.

metratec-mkoehler replied:
Well... but it can't (or at least: won't). It will always return 0 because it will always take zero chars. That's not accepting a new char.

tmon-nordic replied:
The suspend logic in CDC ACM class implementation itself is not quite right.

Linux will suspend the ttyACM device if device claims to support remote wakeup and there is no data from host to device for some time. The device will be woken up by the host if and only if the host wants to send something. If the device has something to send to host, it is up to device to issue to remote wakeup - and it is what Linux expects the ttyACM device supporting remote wakeup will do.

jfischer-no replied:
https://github.com/zephyrproject-rtos/zephyr/pull/73828/commits/490bd42693d5fdd2fa7bfd7c00c34fa076b05e4d#r1646202774 not addressed.

jfischer-no commented on subsys/usb/device/usb_transfer.c:
```diff
@@ -251,7 +251,8 @@ int usb_transfer(uint8_t ep, uint8_t *data, size_t dlen, unsigned int flags,
 
```
Please remove "Partially fixes issue https://github.com/zephyrproject-rtos/zephyr/issues/72786" from the commit message.

>Sometimes loosing USB transfers

Why? k_work_submit_to_queue() can fail for various reasons, what happened in your case?

>Tests with echo device

What is "echo device"?

There is still no description of how to reproduce the bug. I doubt that changes in the sample code alone will be enough to reproduce it. Something else has changed in the Zephyr code.

metratec-mkoehler replied:
> Please remove "Partially fixes issue https://github.com/zephyrproject-rtos/zephyr/issues/72786" from the commit message.

But the 2 commits (formerly one) DO fix the issue. But maybe only adding it to the second commit is fine, too.

metratec-mkoehler replied:
> Why? k_work_submit_to_queue() can fail for various reasons, what happened in your case?

usb_transfer() returns zero (success) even though the queueing failed for one of various reasons. This causes tx_work_handler() to silently drop data (which is the issue described in #72786)

metratec-mkoehler replied:
> What is "echo device"?

A board with samples/subsys/usb/cdc_acm firmware (echoing data to the same USB CDC ACM port)

metratec-mkoehler replied:
> There is still no description of how to reproduce the bug. I doubt that changes in the sample code alone will be enough to reproduce it. Something else has changed in the Zephyr code.

I really don't care so much if you doubt it. The data gets lost, I can see it on wireshark, I can find the reason with a debugger, I can reproduce it (and no, not in 5 minutes, as it's random and was about 1 byte / GB in my case) and the changes fix it. 

Also the description shows the use of functions was just wrong (ignoring return values and states) and should be fixed anyway. With my issue or without it. I am happy to improve my commits as needed and did already add quite some work to this fix but it's time to finish.

jfischer-no replied:
> > Please remove "Partially fixes issue #72786" from the commit message.
> 
> But the 2 commits (formerly one) DO fix the issue. But maybe only adding it to the second commit is fine, too.

You can add Fixes #72786 to PR description, but "Partially fixes issue #72786" should be removed from commit message.

> > Why? k_work_submit_to_queue() can fail for various reasons, what happened in your case?
> 
> usb_transfer() returns zero (success) even though the queueing failed for one of various reasons. This causes tx_work_handler() to silently drop data (which is the issue described in #72786)

You still do not provide a description for why k_work_submit_to_queue() failed.

> > There is still no description of how to reproduce the bug. I doubt that changes in the sample code alone will be enough to reproduce it. Something else has changed in the Zephyr code.
> 
> I really don't care so much if you doubt it. The data gets lost, I can see it on wireshark, I can find the reason with a debugger, I can reproduce it (and no, not in 5 minutes, as it's random and was about 1 byte / GB in my case) and the changes fix it.

But you should care. Exact steps and environment to reproduce it is still missing tought it was requested few times. The reason why k_work_submit_to_queue() sometimes fails is unknown.

metratec-mkoehler replied:
> You still do not provide a description for why k_work_submit_to_queue() failed.

Ah, I didn't get this was even the question, I thought it was about what happens after the error response, not before. My bad.

 I don't know what the error code is but it's very likely -EBUSY (as the other ones are for uninit device, and it's working for the overwhelming number of packets). And I don't see why it is relevant to the fix. On the programmer side I understand the need to go deeper into the root cause, but getting busy response from the queue doesn't seem to be something to bother at a rate of less than 1 per billion tries. Is it that unexpected?

At this low error rate finding the issue would take days or weeks without any real win. There can be a new issue to investigate, but for the USB data loss this is the fix and it took already quite some time to find the issues.

jfischer-no replied:
> I don't know what the error code is but it's very likely -EBUSY (as the other ones are for uninit device, and it's working for the overwhelming number of packets). And I don't see why it is relevant to the fix. On the programmer side I understand the need to go deeper into the root cause, but getting busy response from the queue doesn't seem to be something to bother at a rate of less than 1 per billion tries. 

It is important to find the reason, I still doubt that these changes fix your problem.[Here](https://github.com/zephyrproject-rtos/zephyr/pull/73828#discussion_r1645546441
) you wrote that you can reliably reproduce it. Why is it difficult now?

>Is it that unexpected?

Yes, it is unexpected that k_work_submit_to_queue() fails. Btw, what work_queue is used in your configuration?
 
> At this low error rate finding the issue would take days or weeks without any real win. There can be a new issue to investigate, but for the USB data loss this is the fix and it took already quite some time to find the issues.

This PR cannot be considered to fix #72786 until the reason why k_work_submit failed is clear.

metratec-mkoehler commented on subsys/usb/device/class/cdc_acm.c:
```diff
@@ -256,6 +256,7 @@ static void tx_work_handler(struct k_work *work)
 	uint8_t ep = cfg->endpoint[ACM_IN_EP_IDX].ep_addr;
```
As I say in the initial commit (which I was forced to split and shorten):

usb_transfer() drops data by not checking the return value of k_work_submit_to_queue()
This causes tx_work_handler() to not send the data because it thinks they are in the work queue now. It also thinks the device is occupied but as there is no work queue item it won't get freed again. 

The stuck after reset comes from:
cdc_acm_irq_tx_ready() did not check configured and suspended states 
-> This caused the device to show as ready even though it was not configured after a host reset
The same can happen for RX so I fixed it with it, it's the same issue of ignored state.

Together these cause the issues described in the issue post. 
My changes changed this.

jfischer-no replied:
> As I say in the initial commit (which I was forced to split and shorten):
> 
> usb_transfer() drops data by not checking the return value of k_work_submit_to_queue() This causes tx_work_handler() to not send the data because it thinks they are in the work queue now. It also thinks the device is occupied but as there is no work queue item it won't get freed again.

Can you please provide k_work_submit_to_queue() return value instead of your own interpretation?

> The stuck after reset comes from: cdc_acm_irq_tx_ready() did not check configured and suspended states -> This caused the device to show as ready even though it was not configured after a host reset The same can happen for RX so I fixed it with it, it's the same issue of ignored state.

But it is okay to fill/read the UART FIFO (ringbuffer) event if USB device is not configured or suspended.

> 
> Together these cause the issues described in the issue post. My changes changed this.

I can help you write a commit message later, after we figure out what caused the issue.

metratec-mkoehler replied:
> I can help you write a commit message later, after we figure out what caused the issue.

That's great, learning to fit better into the zephyr workflow will be helpful.  :)
It may delay because today is my last day before the holidays but with some luck it may be done before

metratec-mkoehler committed:
usb: device: Check USB transfer workqueue return value

Sometimes loosing USB transfers

usb_transfer(): k_work_submit_to_queue() return value was ignored
so calling function tx_work_handler() dropped data even though it was
not put to workqueue

Tested with CDC ACM where another fix depends on this return value

Signed-off-by: Martin Koehler <koehler@metratec.com>

metratec-mkoehler commented:
I updated the branch according to requests, restructured the commits and pushed that. 
I am back from my vacation and would like to finalize the pull request.

> Please remove "Partially fixes issue https://github.com/zephyrproject-rtos/zephyr/issues/72786" from the commit message.
Is done by replacing it with "Fixes issue #72786" in the second commit (which does fix it but still depends on the fixed usb_transfer() return value from the first commit)

I will also split the connection check changes from the data loss fix into different commits. 

I would like them to stay in one pull request for convenience and because they got tested together (and I just won't have the time for another test line as the product with the fixed usb code got already shipped), but I can understand if that's not possible. If needed I would split another branch and seperate them from each other. As long as both get pulled this should make no difference.

metratec-mkoehler committed:
usb: cdc_acm: Silent data loss

Fixes issue #72786

What happens:
Packet loss happens on cdc_acm transmission
(below 1 per 1000 packets in my high data load tests),
May depend on usb load, usb speed (host), other queue use.

Test environment:
- Custom board from product development
- RP2040
- USB was handled with the usb_cdc_acm sample as base
- Host PC uses XUbuntu Linux and Python with pyserial
- Test 1: Device echoes data, host PC sends random
  or sorted (counting up) data in diffent block sizes.
  This test showed data missing.
- Test 2: To test if USB RX or TX are the cause the device
  now generates data, usually just a 256 bytes block
  with the index value (0..255) but other data and
  block sizes were also tested. All showed rare data drops
  (in multiples of 8 bytes, single digit bytes per GB of
  transfered data)

Found root issue:
tx_work_handler() does not check usb_transfer() return
value. It drops the data even if not successful.
tx_work_handler() also puts dev->data->tx_ready to
false even if usb_transfer() failed.

Side note:
The tx then gets stuck without cdc_acm_write_cb().
Until then cdc_acm_irq_tx_ready() will always
return false preventing any other transmission until
cdc_acm_write_cb() gets called which can only happen
after cdc_acm_reset_port() and a new configure
(state USB_DC_CONFIGURED in cdc_acm_do_cb())
as far as I see. This issue didn't show in the tests,
I assume the host reconfigures but had no time to check.

Result:
Tests showed the issue was gone after the changes.
Tests were mostly done together with a needed fix to
usb_transfer() as its return value was incorrect
and with a parallel developed change to host reconfigure.
The load tests that showed the issue were used for this, too.

Signed-off-by: Martin Koehler <koehler@metratec.com>

metratec-mkoehler committed:
usb: cdc_acm: Additional connection state checks

What happens:
Restarting the host device caused usb cdc to get stuck.

Test environment:
- Custom board from product development
- RP2040
- USB was handled with the usb_cdc_acm sample as base
  but the sample itself should do
- Host PC uses XUbuntu Linux and Python with pyusb

Test (shortened, code in pull request if needed):
  import usb.core
  devs = usb.core.find(...) #finding all test devices
  for dev in devs:
    dev.reset()

The test showed that nearly every reset caused the device to get stuck
(which was a lot easier to test than restarting the host)

Root cause(s):
- cdc_acm_irq_tx_ready() did not check configured and suspended states
- cdc_acm_irq_rx_ready() did not check if irq is enabled
- cdc_acm_irq_is_pending() had the same issues as its just a test on
  cdc_acm_irq_rx_ready() or cdc_acm_irq_tx_ready()

Tests with both pyusb reset script and a 1 day of
host PC (NUC, Windows) reset loop test showed this fixed the issue
The test included also the fix to the data loss issue but they should
not be connected.

Signed-off-by: Martin Koehler <koehler@metratec.com>

jfischer-no commented on subsys/usb/device/class/cdc_acm.c:
```diff
@@ -288,10 +287,15 @@ static void tx_work_handler(struct k_work *work)
 
-	usb_transfer(ep, data, len, USB_TRANS_WRITE,
-		     cdc_acm_write_cb, dev_data);
+	ret = usb_transfer(ep, data, len, USB_TRANS_WRITE,
+			   cdc_acm_write_cb, dev_data);
+	if (ret != 0) {
+		(void)ring_buf_get_finish(dev_data->tx_ringbuf, 0);
+		return;
+	}
```
If usb_transfer() fails, it disables TX interrupts until bus reset or similar. Instead, error branch should reschedule tx_work with some delay, maybe 1ms.

jfischer-no reviewed (changes requested):
Previous comments are still not addressed.

carlescufi commented:
@metratec-mkoehler could you please go over @jfischer-no's review comments and address them? Then we may have a chance to get this merged in time for 3.7.0.

metratec-mkoehler commented:
I talked to my manager and got a go to prepare a test case to support this pull request. 
I will start now and hopefully it's possible to reproduce on my nrf dev board.
I will show the test projects and results asap-

metratec-mkoehler commented:
I have big issues to reproduce the data loss while logging or using other tools. And it is already annoyingly rare. So I decided to split the reset issue from this pull request. I will make an issue today or tommorow, it's easy to reproduce and the fix should be working without the other changes. 

This way we can have at least one fix in (and it's easier to talk about errors at this point because I have no expertise on the USB protocol itself).

I will add more info on the data loss issue as soon as possible, too.

mmahadevan108 commented:
@metratec-mkoehler, do you have any updates for this PR.

metratec-mkoehler commented:
Sadly I don't. 

The changes were partially moved into different pull requests 
[https://github.com/zephyrproject-rtos/zephyr/pull/76171] (Host reset)
[https://github.com/zephyrproject-rtos/zephyr/pull/76173] (Logging error, already merged)

I can't reproduce the data loss issue. It is possible the loss was not in Zephyr or, if it was, the changes made in this PR only randomly (race condition?) fixed the issue for me. Other changes, not related to the driver, also "fixed" them in my case. This is highly dissatisfying but it's what it is. 

I will mark this PR as closed (the host reset fix is still pending in the other PR)
